servo/servo
# DevTools: Implement watcher actor

The new DevTools protocol creates the figure of the [Watcher actor](https://firefox-source-docs.mozilla.org/devtools/backend/actor-hierarchy.html). This is a mediator that is assigned to an element (tab/browsing context, process, worker...) and shows DevTools what resources and targets are available to inspect. Having one is the first step into restoring the full tab inspection functionality, which will be very useful since it will allow to see the source code of a website, the console logs, the render timeline, etc...

The new watcher actor is able to register itself and respond with the available targets and resources. This makes it so the DevTools UI on Firefox's side appears again, though it doesn't have full functionality yet.

<details>

<summary><b>Messages sent when clicking Inspect</b></summary>

<- (Servo receives, Firefox Sends) | -> (Servo sends, Firefox receives)

- [x] <- **getWatcher** _request a watcher actor for the tab we want to inspect_
- [x] -> **actor: watcher, traits** _replies with the watcher actor and the resources and targets it supports, for example, it can support debugging the console but not the cookies_
- [x] <- **watchTargets, targetType: frame** _tell the watcher actor that we want to inspect the frame of the tab_
- [x] -> **target-available-form, target: browsing context** _it replies with the browsing context of the tab, and all of the related debugging actors' ids, for example the console actor or the inspector actor_
- [x] -> **frameUpdate, frames**
- [x] -> ACK
- [x] <- **watchResources, resourceTypes: document-event** _the first resource we ask for is the document information itself_
- [x] -> **resource-available-form, resources** _reply with the document information, such as the url and the title_
- [x] -> ACK
- [x] <- **getTargetConfigurationActor** / **getServerConfigurationActor** 

</details>

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes are a part of #32404
- [x] These changes do not require tests because there are no tests yet for DevTools.

### Manual testing

1. Compile and run Servo with devtools enabled. You can also enable debug logs for the devtools module so that you can see the JSON messages being sent.

```sh
RUST_LOG="error,devtools=debug" ./mach run --devtools=1234
```

2. Open a recent version of Firefox and go to `about:debugging`. In Setup, add `localhost:1234` (or the port you selected before) as a network location.

![image](https://github.com/servo/servo/assets/22449369/91d87705-211e-40c5-9847-1175b8568a30)

3. On the sidebar, click on Connect next to `localhost:1234`. Now go to the terminal where you are running Servo and press "y" to accept the incoming connection.

![image](https://github.com/servo/servo/assets/22449369/79f15570-a261-460f-adf8-d74799f6a28e)

4. Go back to Firefox and click on `localhost:1234` on the sidebar. You should be presented with a screen like this:

![image](https://github.com/servo/servo/assets/22449369/b39930b6-10ca-4a26-90ad-500fd1432c2f)

5. Choose one tab and click Inspect. A new window should open with an inspector for the tab. Right now the only feature that is working to some extent is the console. You can't see any message yet, but you can run some javascript. For example, writing `location.reload()` should reload the page. Or, on the https://servo.org website, use `document.getElementsByClassName("hero-body")[0].style.backgroundColor = "royalblue" ` to change the background color:

![image](https://github.com/servo/servo/assets/22449369/c8febc8a-d99c-40d2-861f-54401b8c29e0)

Refer to [this comment](https://github.com/servo/servo/issues/32404#issuecomment-2168414796) for more information on how to test the protocol.
